src/tpm2_pytss/policy.py

```diff
@@ -0,0 +1,594 @@
+# SPDX-License-Identifier: BSD-2
+from .internal.utils import _lib_version_atleast, _chkrc
+
+if not _lib_version_atleast("tss2-policy", "3.2.0-63-gdcdc8412"):
+    raise NotImplementedError(
+        "tss2-policy not installed or version is less then 3.2.0-63-gdcdc8412"
+    )
+
+from .types import (
+    TPM2B_DIGEST,
+    TPMS_NV_PUBLIC,
+    TPM2B_NAME,
+    TPMT_PUBLIC,
+    TPM2B_NONCE,
+    TSS2_OBJECT,
+    TSS2_POLICY_PCR_SELECTION,
+    TPM2_HANDLE,
+)
+from .constants import TPM2_ALG, ESYS_TR, TSS2_RC
+from .TSS2_Exception import TSS2_Exception
+from ._libtpm2_pytss import ffi, lib
+from .ESAPI import ESAPI
+from enum import Enum
+from typing import Callable, Union
+
+
+class policy_cb_types(Enum):
+    """Policy callback types"""
+
+    CALC_PCR = 0
+    CALC_NAME = 1
+    CALC_PUBLIC = 2
+    CALC_NVPUBLIC = 3
+    EXEC_AUTH = 4
+    EXEC_POLSEL = 5
+    EXEC_SIGN = 6
+    EXEC_POLAUTH = 7
+    EXEC_POLAUTHNV = 8
+    EXEC_POLDUP = 9
+    EXEC_POLACTION = 10
+
+
+@ffi.def_extern()
+def _policy_cb_calc_pcr(selection, out_selection, out_digest, userdata):
+    """Callback wrapper for policy PCR calculations
+
+    Args:
+        selection (struct TSS2_POLICY_PCR_SELECTION): in
+        out_selection (struct TPML_PCR_SELECTION): out
+        out_digest (struct TPML_DIGEST): out
+        userdata (ffi handle): in
+    """
+    pi = ffi.from_handle(userdata)
+    cb = pi._get_callback(policy_cb_types.CALC_PCR)
+    if not cb:
+        return TSS2_RC.POLICY_RC_NULL_CALLBACK
+    try:
+        selcopy = ffi.new("TSS2_POLICY_PCR_SELECTION *", selection[0])
+        sel = TSS2_POLICY_PCR_SELECTION(_cdata=selcopy)
+        cb_selection, cb_digest = cb(sel)
+        out_selection.count = cb_selection.count
+        for i in range(0, cb_selection.count):
+            out_selection.pcrSelections[i] = cb_selection[i]._cdata
+        out_digest.count = cb_digest.count
+        for i in range(0, cb_digest.count):
+            out_digest.digests[i].buffer = cb_digest[i]
+            out_digest.digests[i].size = len(cb_digest[i])
+    except Exception as e:
+        rc = (
+            e.rc if isinstance(e, TSS2_Exception) else TSS2_RC.POLICY_RC_GENERAL_FAILURE
+        )
+        pi._callback_exception = e
+        return rc
+    return 0
```

Comment:
`TSS2_RC.SUCCESS` or whatever it is.

The logic through here took me a minute, but really well thought out kudos. The Python callback can raise an exception, which gets translated to an RC which is returned through the callback code and caught in checkrc in the execute or whatever invoking call and then re-raised there.

Reply:
Yeah, that worked out well :-)
Might be worth looking into doing a similar callback setup with FAPI

Reply:
> Yeah, that worked out well :-) Might be worth looking into doing a similar callback setup with FAPI

Oh how did they do it? I didn't really follow the FAPI stuff closely as I rarely use FAPI.

Reply:
From my understanding of the FAPI code: prepare_headers generates a list of ten extern functions for the each callback type, when registering a callback, it finds the first unlocked callback in a list, locks in and then uses that when setting the callback for the C API